src/H5FDsubfiling/H5FDsubfiling.c

```diff
@@ -2581,138 +2043,913 @@ H5FD__subfiling_ctl(H5FD_t *_file, uint64_t op_code, uint64_t flags, const void
+         * for any particular subfile; e.g. the maximum number of
+         * I/O requests for any particular subfile
+         */
+        row_stripe_idx_start = stripe_idx - first_subfile;
+        row_stripe_idx_final = final_stripe_idx - last_subfile;
+        cur_iovec_depth      = (size_t)((row_stripe_idx_final - row_stripe_idx_start) / num_subfiles) + 1;
+
+        /*
+         * If the I/O request "wrapped around" and ends in a subfile
+         * less than the subfile we started in, subtract one from the
+         * I/O vector length to account for "empty space". This can be
+         * visualized as follows:
+         *
+         *   SUBFILE 0   SUBFILE 1   SUBFILE 2   SUBFILE 3
+         *  _______________________________________________
+         * |           |           |   XXXXX   |   XXXXX   | ROW 0
+         * |   XXXXX   |   XXXXX   |   XXXXX   |   XXXXX   | ROW 1
+         * |   XXXXX   |   XXXXX   |           |           | ROW 2
+         * |           |           |           |           | ROW ...
+         * |           |           |           |           |
+         * |           |           |           |           |
+         * |           |           |           |           |
+         * |___________|___________|___________|___________|
+         *
+         * Here, `stripe_idx` would be calculated as 2 (I/O begins in
+         * the 3rd stripe, or subfile index 2), `first_subfile` would be
+         * calculated as 2 and the starting "row" (row_stripe_idx_start)
+         * would be calculated as "row" index 0. `final_stripe_idx` would
+         * be calculated as 9, `last_subfile` would be calculated as
+         * (9 % 4) = 1 and the ending "row" (row_stripe_idx_final) would
+         * be calculated as (9 - 1) = 8. Thus, the calculated I/O vector
+         * length would be ((8 - 0) / 4) + 1 = 3. However, since there is
+         * no I/O to stripe indices 0 and 1 (residing in "row" 0 of subfile
+         * index 0 and 1, respectively), it can be seen that the real I/O
+         * vector length is 2.
+         */
+        if (last_subfile < first_subfile)
+            cur_iovec_depth--;
+
+        loc_max_iovec_depth = MAX(cur_iovec_depth, loc_max_iovec_depth);
+
+        /*
+         * Determine the maximum number of subfiles this piece of the
+         * I/O request could touch
+         */
+        if (data_size >= block_size) {
+            /*
+             * I/O of a size greater than the block size definitionally
+             * touches all subfiles at least once.
+             */
+            cur_max_num_subfiles = (size_t)num_subfiles;
+        }
+        else if (data_size < stripe_size) {
+            /*
+             * I/O of a size smaller than the stripe size could
+             * touch one or two subfiles at most, depending on
+             * the file offset.
+             */
+            cur_max_num_subfiles = 2;
+        }
+        else {
+            cur_max_num_subfiles = data_size / stripe_size;
```

Comment:
I'm not sure this calculation is correct. For example if the block starts at a strip boundary, the strip size is 1000, and the data size is 2001, this will return 2 subfiles, when it really touches 3. And I think it needs to take into account the actual starting position, not just whether it's on a boundary, to avoid wasting space.

I think something like `(((cur_file_offset % stripe_size) + data_size - 1) / stripe_size) + 1` could work

Reply:
The below `if (cur_file_offset % stripe_size)` is responsible for adding 1 to the value when the offset is aligned to a stripe boundary. I was less concerned about wasting space since an extra entry in the arrays would make very little difference.

Reply:
In the case above this code results in an array that's too small

Reply:
I see that now. I couldn't immediately find a case where the above suggestion is problematic, so I'll switch to that instead.